--- old/util.c	2011-05-11 13:33:05.000000000 -0700
+++ new/util.c	2011-05-11 13:36:31.000000000 -0700
@@ -90,7 +90,20 @@
 		unsigned char c = *str;
 		if ((c == '\r') || (c == '\n') ||
 			((c < 32) && (c != '\t')))
-			return false;
+		  {
+      if ( c == '\r' ) {
+        fprintf( stderr, "uci: illegal character: '\r'\n" );
+        return false;
+      }
+      if ( c == '\n' ) {
+        fprintf( stderr, "uci: illegal character: '\n'\n" );
+        return false;
+      }
+      if ( (c < 32) && (c!='\t') ) {
+        fprintf( stderr, "uci: illegal character: '%c' (%d)\n", c, c );
+        return false;
+      }
+		  }
 		str++;
 	}
 	return true;
