diff -Naur linux-2.6.36/drivers/net/sky2.c linux-2.6.36-new/drivers/net/sky2.c
--- linux-2.6.36/drivers/net/sky2.c	2010-10-20 13:30:22.000000000 -0700
+++ linux-2.6.36-new/drivers/net/sky2.c	2011-09-04 11:08:16.000000000 -0700
@@ -4855,6 +4855,9 @@
 	pci_set_drvdata(pdev, hw);
 	pdev->d3_delay = 150;
 
+	//debug_Aaron
+        pcie_set_readrq(pdev, 512);
+
 	return 0;
 
 err_out_unregister:
diff -Naur linux-2.6.36/drivers/net/tg3.c linux-2.6.36-new/drivers/net/tg3.c
--- linux-2.6.36/drivers/net/tg3.c	2010-10-20 13:30:22.000000000 -0700
+++ linux-2.6.36-new/drivers/net/tg3.c	2011-09-04 11:08:16.000000000 -0700
@@ -6831,7 +6831,9 @@
 
 	if (GET_ASIC_REV(tp->pci_chip_rev_id) != ASIC_REV_5785) {
 		if (tp->tg3_flags2 & TG3_FLG2_PCI_EXPRESS)
-			pcie_set_readrq(tp->pdev, 4096);
+			//debug_Aaron
+			//pcie_set_readrq(tp->pdev, 4096);
+			pcie_set_readrq(tp->pdev, 512);
 		else {
 			pci_write_config_byte(tp->pdev, PCI_CACHE_LINE_SIZE,
 					      tp->pci_cacheline_sz);
@@ -7025,7 +7027,9 @@
 				      tp->pcie_cap + PCI_EXP_DEVCTL,
 				      val16);
 
-		pcie_set_readrq(tp->pdev, 4096);
+		//debug_Aaron
+		//pcie_set_readrq(tp->pdev, 4096);
+		pcie_set_readrq(tp->pdev, 512);
 
 		/* Clear error status */
 		pci_write_config_word(tp->pdev,
@@ -13109,8 +13113,10 @@
 		u16 lnkctl;
 
 		tp->tg3_flags2 |= TG3_FLG2_PCI_EXPRESS;
-
-		pcie_set_readrq(tp->pdev, 4096);
+		
+		//debug_Aaron
+		//pcie_set_readrq(tp->pdev, 4096);
+		pcie_set_readrq(tp->pdev, 512);
 
 		pci_read_config_word(tp->pdev,
 				     tp->pcie_cap + PCI_EXP_LNKCTL,
